test/api.jl

```diff
@@ -115,5 +116,39 @@ recommendation = provide_recommendation(trial_vector)
 recommendation_multi = provide_recommendation(trial_vector_multihyperparams)
 @test recommendation_multi == Dict(:x => 3, :y => 7)
 
+# Test case reproducing the issue in “obs_memo” function from the higher level
+# More details on this issue see https://github.com/IQVIA-ML/TreeParzen.jl/issues/86
+a_list = [8, 9, 10]
+b_list = [5, 6, 7]
+space = Dict(
+    :a => HP.Choice(:a, a_list),
+    :b => HP.Choice(:b, b_list)
+    )
+trials = TreeParzen.Trials.Trial[]
+
+for i in 1:100
+    trial = ask(space, trials, TreeParzen.Config())
+    tell!(trials, trial, trial.hyperparams[:b]/trial.hyperparams[:a])
+end
+# With this example space and loss function, expected vals for the best param should be different
+# but with the issue in “obs_memo” function, they are equal due to same nid
+equal_vals = []
+for i in 21:100
+    push!(equal_vals, trials[i].vals[:a] == trials[i].vals[:b] ? 1 : 0)
+end
+equal_vals_percentage = mean(equal_vals) * 100
+# the equal_vals_percentage should be smaller than 30 %
+@test_broken equal_vals_percentage < 30
+
+# To have the smaller loss, the expected best hyperparams[:a] is larger than a_list[2]
+# and hyperparams[:b] is smaller than b_list[2]
+samples_a = getindex.(getfield.(trials, Ref(:hyperparams)), Ref(:a))
+samples_b = getindex.(getfield.(trials, Ref(:hyperparams)), Ref(:b))
+expected_mean_a = mean(samples_a[21:end])
+expected_mean_b = mean(samples_b[21:end])
+
+# As :a is checked first expected_mean_a > a_list[2] is True but after the bug is fixed
+# expected_mean_b < b_list[2] should be True too
+@test_broken (expected_mean_a > a_list[2]) && (expected_mean_b < b_list[2])
```

Comment:
```suggestion
a_list = [10, 14, 19]
b_list = [1, 4, 9]

space = Dict(
    :a => HP.Choice(:a, a_list),
    :b => HP.Choice(:b, b_list),
)

n_samples = 100
n_random = 100

config = TreeParzen.Config(;linear_forgetting=n_samples + n_random, draws=50, threshold=0.1, random_trials=n_random)
posterior_start = config.random_trials + 1

trials = TreeParzen.Trials.Trial[]

for i in 1:n_samples+n_random
    trial = ask(space, trials, config)
    tell!(trials, trial, trial.hyperparams[:b] / trial.hyperparams[:a])
end

posterior_trials = trials[posterior_start:end]

samples_a = getindex.(getfield.(posterior_trials, Ref(:hyperparams)), Ref(:a))
samples_b = getindex.(getfield.(posterior_trials, Ref(:hyperparams)), Ref(:b))

vals_a = getindex.(getfield.(posterior_trials, Ref(:vals)), Ref(:a))
vals_b = getindex.(getfield.(posterior_trials, Ref(:vals)), Ref(:b))

same_indices_pct = mean(vals_a .== vals_b) * 100
# With this example space and loss function, expected vals for the best param should be mostly different
# but with the issue in “obs_memo” function, they are equal due to same nid
@test_broken same_indices_pct <= 10

# To have the smaller loss, the expected best hyperparams[:a] is the largest value of a
# and expected best hyperparams[:b] is the smallest value of b
expected_a = a_list[end]
expected_b = b_list[1]

# values for b end up meeting expectations because that is the wrong posterior we get
# but the posterior for a is being overwritten while the bug is present
@test_broken (mean(samples_b .== expected_b) * 100) >= 50 &&  (mean(samples_a .== expected_a) * 100) >= 50
```

I hope you don't mind me taking the liberty to rewrite this a bit -- I wanted to increase the clarity but also reduce the chances of it failing by chance.

I've set the values for the sets to draw from and the hyperparameter optimiser parameters (config) to make it easy to converge to the best loss and make the optimiser more greedy.

This helps to reduce to chance for random failures.

With the existing version of code I ran for 10's of thousands of times and it would fail quite often, but with this version I ran it for 10's of thousands of times and it didn't fail.